src/Microsoft.IdentityModel.Xml/DsigSerializer.cs

```diff
@@ -101,58 +100,22 @@ public virtual KeyInfo ReadKeyInfo(XmlReader reader)
-                            }
-                            else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509IssuerSerial, XmlSignatureConstants.Namespace))
-                            {
-                                if (keyInfo.SerialNumber != null)
-                                    throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509IssuerSerial);
-
-                                ReadIssuerSerial(reader, keyInfo);
-                            }
-                            else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509SKI, XmlSignatureConstants.Namespace))
-                            {
-                                if (keyInfo.SKI != null)
-                                    throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509SKI);
-
-                                ReadSKI(reader, keyInfo);
-                            }
-                            else if (reader.IsStartElement(XmlSignatureConstants.Elements.X509SubjectName, XmlSignatureConstants.Namespace))
-                            {
-                                if (keyInfo.SubjectName != null)
-                                    throw XmlUtil.LogReadException(LogMessages.IDX30015, XmlSignatureConstants.Elements.X509SubjectName);
-
-                                ReadSubjectName(reader, keyInfo);
-                            }
-                            else
-                            {
-                                // Skip the element since it is not one of  <X509Certificate>, <X509IssuerSerial>, <X509SKI>, <X509SubjectName>
-                                Logger.WriteWarning(LogMessages.IDX30300, reader.ReadOuterXml());
-                            }
-                        }
-
-                        // </X509Data>
-                        reader.ReadEndElement();
+                        keyInfo.X509Data.Add(ReadX509Data(reader));
                     }
+
                     // <RetrievalMethod>
                     else if (reader.IsStartElement(XmlSignatureConstants.Elements.RetrievalMethod, XmlSignatureConstants.Namespace))
                     {
                         keyInfo.RetrievalMethodUri = reader.GetAttribute(XmlSignatureConstants.Attributes.URI);
                         reader.ReadOuterXml();
                     }
-                    // TODO <KeyName>
-                    //else if (reader.IsStartElement(XmlSignatureConstants.Elements.KeyName, XmlSignatureConstants.Namespace))
-                    //{
-                    //}
+
+                    // <KeyName>
+                    else if (reader.IsStartElement(XmlSignatureConstants.Elements.KeyName, XmlSignatureConstants.Namespace))
+                    {
+                        keyInfo.KeyName = reader.ReadElementContentAsString(XmlSignatureConstants.Elements.KeyName, XmlSignatureConstants.Namespace);
+                    }
+
```

Comment:
We don't need a new line after the else if, only at the end of a clause.
Sorry if that wasn't clear.

Reply:
Done!